builtin/process_osh.py

```diff
@@ -450,28 +458,116 @@ def Run(self, cmd_val):
         argv = cmd_val.argv[1:]
         if len(argv) == 0:
             # umask() has a dumb API: you can't get it without modifying it first!
+            # see: https://man7.org/linux/man-pages/man2/umask.2.html
             # NOTE: dash disables interrupts around the two umask() calls, but that
             # shouldn't be a concern for us.  Signal handlers won't call umask().
             mask = posix.umask(0)
-            posix.umask(mask)  #
+            posix.umask(mask)
             print('0%03o' % mask)  # octal format
             return 0
 
         if len(argv) == 1:
-            a = argv[0]
-            try:
-                new_mask = int(a, 8)
-            except ValueError:
-                # NOTE: This also happens when we have '8' or '9' in the input.
-                print_stderr(
-                    "oils warning: umask with symbolic input isn't implemented"
-                )
-                return 1
-
-            posix.umask(new_mask)
-            return 0
+            new_mask = 0  # type: int
+            first_arg = argv[0]
+            if first_arg[0].isdigit():
+                try:
+                    new_mask = int(first_arg, 8)
+                except ValueError:
+                    # NOTE: This also happens when we have '8' or '9' in the input.
+                    print_stderr("oils warning: `%s` is not an octal number" %
+                                 first_arg)
+                    return 1
+
+                posix.umask(new_mask)
+                return 0
+            elif first_arg[0] in "ugoa":
+                # TODO: it's possible to avoid this extra syscall in cases where we don't care about
+                # the initial value (ex: umask u=rwx,a=rwx) although it's non-trivial to determine
+                # when, so it's probably not worth it
+                initial_mask = posix.umask(0)
+                new_mask = initial_mask
+
+                for arg_component in first_arg.split(","):
+                    maybe_new_mask = self._SymbolicToOctal(
+                        new_mask, arg_component)
+                    if maybe_new_mask is None:
+                        posix.umask(initial_mask)
+                        return 1
+
+                    new_mask = maybe_new_mask.value
+
+                posix.umask(new_mask)
+                return 0
+
+            print_stderr(
+                "oils warning: `%s` is an invalid symbolic mode operator" %
+                first_arg[0])
+            return 1
 
-        e_usage('umask: unexpected arguments', loc.Missing)
+        e_usage("unexpected number of arguments", loc.Missing)
+
+    def _SymbolicToOctal(self, initial_mask, arg_component):
+        # type: (int, str) -> Optional[Int]
```

Comment:
We often use regular `int` and then something like `BAD_MASK = -1`  

Or you can also use `ok, octal = _ParseOctal()`, sort of like Go

I think one of those will work here